network/proxy/mucp/mucp.go

```diff
@@ -79,155 +84,126 @@ func readLoop(r server.Request, s client.Stream) error {
 	}
 }
 
-func (p *Proxy) getRoute(service string) ([]string, error) {
-	// converts routes to just addresses
-	toNodes := func(routes []table.Route) []string {
-		var nodes []string
-		for _, node := range routes {
-			address := node.Address
-			if len(node.Gateway) > 0 {
-				address = node.Gateway
-			}
-			nodes = append(nodes, address)
+// toNodes returns a list of node addresses from given routes
+func toNodes(routes map[uint64]table.Route) []string {
+	var nodes []string
+	for _, node := range routes {
+		address := node.Address
+		if len(node.Gateway) > 0 {
+			address = node.Gateway
 		}
-		return nodes
+		nodes = append(nodes, address)
 	}
+	return nodes
+}
 
+func (p *Proxy) getRoute(service string) ([]string, error) {
 	// lookup the route cache first
 	p.RLock()
 	routes, ok := p.Routes[service]
-	// got it!
 	if ok {
-		p.RUnlock()
 		return toNodes(routes), nil
 	}
+	p.Routes[service] = make(map[uint64]table.Route)
```

Comment:
This is being created while using a read lock so I think it introduces some sort of race

Reply:
Oh, amazing catch! I'll blame this one on my dyslexia 😜